working/augmentation-libraries/parts_with_imports.md

```diff
@@ -0,0 +1,544 @@
+associated *language version*. If any file has a language-version override
+marker (a line like `// @dart=3.12` before any Dart code), then *every file* in
+the library *must* have a language override marker. _(And they must still have
+the same language version, so it must be the same marker.)_
+
+Also, every file in a library must belong to the same *package*. The Dart
+language itself has no notion of packages, but the tooling uses a file’s
+package to derive its default language version. The Dart SDK will require that
+all files in a package belong to the same library, ensuring that they’ll always
+have the same language version. Also, if the library file is inside the `lib/`
+directory, so it has a `package:` URI as canonical URI, then so must all its
+sub-part files. We haven’t specified these requirements before,
+it has always been assumed, but technically it is possible to write programs
+where a part belongs to different package than their library. The Dart SDK’s
+multi-language-version support, which based on files belonging to packages,
+will not support libraries that are not entirely in a single package.
+
+That is, the extra restrictions enforced by Dart tools are:
+
+*   If any file of a library has a `// @dart=` language version marker, then
+    *all* files in that library must have a language version marker with
+    *the same* language version.
+    Or, phrased differently starting at the root:
+    *   If a library file has a language version marker,
+        then it’s a compile-time error for any sub-part file
+        to not have the same language version marker.
+    *   If a library file has no language version marker,
+        then it’s a compile-time error for any sub-part file
+        to have a language version marker.
+
+*   If a library file belongs to a package,
+    then all its sub-part files must belong to the same package.
+*   If a library file has a `package:` URI (is located inside the package-URI
+    directory specified by the package configuration, usually
+    `"packageUri": "lib/"` of a `package_config.json` file),
+    then all its sub-part files must also be inside that package-URI directory.
+    *   _A library file outside of the `lib/` directory can technically have
+        part files inside the `lib/` directory, specified using `package:`
+        URIs, but those part files can only be included in a program by
+        referencing the library file, which means the part files might as well
+        be placed outside of the `lib/` directory._
+
+### User guidance tooling
+
+The analyzer and analysis server needs to support and understand the new
+feature. No new user-facing features are needed, but some error handling
+and user guidance may be useful. The following are ideas for
+hypothetical features that the tool may choose to add.
+
+##### Annotations applying to sub-tree
+
+Usually an annotation placed on the `library` declaration applies to the entire
```

Comment:
This might have some implications for macros... but I think it is fine. We could always look at allowing macros on parts later on if we have a use case.